extension/encoding/jsonlogencodingextension/extension.go

```diff
@@ -78,35 +138,43 @@ func (e *jsonLogExtension) Shutdown(_ context.Context) error {
 	return nil
 }
 
-func (e *jsonLogExtension) logProcessor(ld plog.Logs) ([]byte, error) {
-	logs := make([]logBody, 0, ld.LogRecordCount())
+// streamReader is a wrapper to process input stream and return processed JSON records one by one
+type streamReader struct {
+	decoder *json.Decoder
+	current map[string]any
+	err     error
+	done    bool
+}
 
-	rls := ld.ResourceLogs()
-	for i := 0; i < rls.Len(); i++ {
-		rl := rls.At(i)
-		resourceAttrs := rl.Resource().Attributes().AsRaw()
+func newStreamReader(r io.Reader) *streamReader {
+	return &streamReader{
+		decoder: json.NewDecoder(r),
+	}
+}
 
-		sls := rl.ScopeLogs()
-		for j := 0; j < sls.Len(); j++ {
-			sl := sls.At(j)
-			logSlice := sl.LogRecords()
-			for k := 0; k < logSlice.Len(); k++ {
-				log := logSlice.At(k)
-				logEvent := logBody{
-					Body:               log.Body().AsRaw(),
-					ResourceAttributes: resourceAttrs,
-					LogAttributes:      log.Attributes().AsRaw(),
-				}
-				logs = append(logs, logEvent)
-			}
+func (r *streamReader) next() bool {
+	if r.done {
+		return false
+	}
+
+	err := r.decoder.Decode(&r.current)
```

Comment:
Decode will merge over the top of an existing map, so `r.current` should be emptied first.